src/Framework/AspNetCoreAnalyzers/src/SourceGenerators/PublicTopLevelProgramGenerator.cs

```diff
@@ -0,0 +1,44 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.AspNetCore.SourceGenerators;
+
+[Generator]
+public class PublicProgramSourceGenerator : IIncrementalGenerator
+{
+    private const string PublicPartialProgramClassSource = """
+// <auto-generated />
+public partial class Program { }
+""";
+
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var internalGeneratedProgramClass = context.CompilationProvider
+            // Get the entry point associated with the compilation, this maps to the Main method definition
+            .Select((compilation, cancellationToken) => compilation.GetEntryPoint(cancellationToken))
```

Comment:
According to [the guidance on writing incremental generators](https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md#authoring-a-cache-friendly-generator) it's better to split things out to multiple `Selects` so that there's an opportunity to cache more frequently in between steps.

Reply:
Surely that's only true for things that change independently?  Retrieving and casting the containing type in the first select doesn't seem like something that could affect caching.

Reply:
Yes, I suppose the first two selects could be collapsed into one given the things they evaluate are closely related but I don't think there's anything _actively_ harmful about splitting up the transformation.